Update 2201.4.0 release note (#6336)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.4.0/RELEASE_NOTE.md

## Overview of Ballerina Swan Lake 2201.4.0

> <em>2201.4.0 (Swan Lake) is the fourth major release of 2022, and it includes a new set of features and significant improvements to the compiler, runtime, standard library, and developer tooling. It is based on the 2022R4 version of the Language Specification.</em> 

Reviewer: It is no longer correct to say `the fourth major release of 2022` as we are in 2023 now

> <em>2201.4.0 (Swan Lake Update 4) is the fourth major release of Ballerina Swan Lake, and it includes a new set of features and significant improvements to the compiler, runtime, standard library, and developer tooling. It is based on the 2022R4 version of the Language Specification.</em> 

Reviewer: Is it correct to say `fourth major release` here? Referring to the word `major` as it has a different meaning in semver. (Seems we had the same term in previous release notes too, but only noticed now)

What if we say `fourth update release` instead. @sameerajayasoma thoughts?

Reviewer 2: `fourth update release` is the correct term.

Author: Updated.

## Install Ballerina

If you have not installed Ballerina, download the [installers](/downloads/#swanlake) to install.

## Language updates

### New features

#### Support for annotations on tuple members

> - Annotations are now supported on tuple members. Tuple members can be annotated using annotations declared with the `field` attachment point.

Reviewer 3: Why are we making this point form? Not required.

```suggestion
Annotations are now supported on tuple members. Tuple members can be annotated using annotations declared with the `field` attachment point.
```

Reviewer 3: Please check other places too.

Reviewer 3: Basically this is content under a heading. We don't need separate bullet points. This is what we did previously also.

  ```ballerina
  annotation annot on field;

  type T [int, @annot
      string ];
  ```

  Annotations are not allowed on the tuple rest descriptor. 

### Bug fixes

- Annotation values of fields of record type descriptors that are not defined with a type definition are now accessible at runtime.

> - To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.4.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.4.0+is%3Aclosed+label%3AType%2FBug).

Reviewer 3: This doesn't have to be a separate bullet point.
```suggestion
To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.4.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.4.0+is%3Aclosed+label%3AType%2FBug).
```

## Runtime updates

### New features

#### Get the strand dump during `bal test`

- When running the tests in a Ballerina package or a file using the `bal test` command, the strand dump can be obtained by sending the `SIGTRAP` signal to that process.

### Improvements

> #### Get the user-defined type name on Singleton types

Reviewer 3: ```suggestion
#### Get the user-defined type name on singleton types
```

Annotations are now supported on tuple members. Tuple members can be annotated using annotations declared with the `field` attachment point.

  ```ballerina
  annotation annot on field;

  type T [int, @annot
      string ];

>   ```

Reviewer: @MaryamZi Is this code formatting correct?

Reviewer 3: No, can be on the  same line here. I believe this was changed when there were bullet points previously.

@praneesha, can you fix the indentation for the entire code block and also the line after. Doesn't have to be indented?

Author: This was what I got after doing `Format document` on the VS Code editor. Will check and fix.

Reviewer 3: It's probably because you were using an older version of Ballerina instead of the RC.

Author: Fixed.

  Annotations are not allowed on the tuple rest descriptor. 

### Bug fixes

> - Annotation values of fields of record type descriptors that are not defined with a type definition are now accessible at runtime.

Reviewer: Why only this bug is special cased?

Reviewer 3: It is in a way a behavioural change (we get annotation values we previously didn't get at runtime), thus we've emphasized it over other bug fixes.

We used to this for bug fixes that are breaking changes.

Reviewer: Is that a breaking change? Then shouldn't we add a topic as breaking changes? @MaryamZi

To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.4.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.4.0+is%3Aclosed+label%3AType%2FBug).

## Runtime updates

### New features

#### Get the strand dump during `bal test`

> - When running the tests in a Ballerina package or a file using the `bal test` command, the strand dump can be obtained by sending the `SIGTRAP` signal to that process.

Reviewer 3: Please recheck all bullet points. (Generally) we don't use bullet points if there is a heading for the new feature/improvement.

Author: Shall we go with the guideline that we add bullets only for a list and if there is only one point, we add it without a bullet point. @anupama-pathirage @MaryamZi

Reviewer: +1 I think we don't need bullet point if it is only one item in the list

### New features

#### `constraint` package

> - Introduced the `@pattern` constraint on string types.
> - Added constraint validation support for `readonly` types.

Reviewer 4: ```suggestion
- Added constraint validation support for `readonly` types.
```

    ```

#### JSON-to-record converter

- Improved the JSON-to-record converter tool to be more context-aware and generate records with non-conflicting names.

#### Language Server

- Improved the completion and code action support for pulled modules.
- Improved sorting of module-level completion items.
- Used the code action resolve request for compiler plugin code actions.

#### OpenAPI Tool

- Added support for the `additionalProperties` attribute in OpenAPI object schemas. With this support, the generated Ballerina records can be either open or closed records based on the `additionalProperties` details.
  >**Info:** Thereby, some of your already generated open records may change into closed records when re-generating using 2201.4.0 (and above) versions.

> - Improved the support for the `nullable:true` property OpenAPI schema to generate record fields with default values (e.g., `string? name = ();`) instead of making the field both nilable and optional (e.g., `string? name?;`).

Reviewer 5: ```suggestion
- Improved the behaviour for the `nullable:true` property in OpenAPI schemas, to generate record fields with default values (e.g., `string? name = ();`), instead of making the field both nilable and optional (e.g., `string? name?;`).
```
